### DISCUSSION TITLE
Can cluster slots be cached beyond request lifecycle?

### DISCUSSION THREAD

**Discussion by gskema** [phpredis/phpredis#2456] [Q&A] [answered]
Hello,
we have ElastiCache cluster, we are using `persistent: true` argument in `RedisCluster` constructor + we have enabled `redis.clusters.cache_slots = 1`
What I noticed is that in every PHP request the first operation to cluster takes twice as long - I assume it's fetch slots + actual operation. The subsequent operations take half the time (only GET operation).

1) Is it intended that slots are cached per-requests, or can they be configured to be cached for whole PHP-FPM process pool for longer time? 
2) Is it an issue with configuration like missing `redis.pconnect.pooling_enabled` (I expected this wouldn't apply to RedisCluster)
3) Does `redis.clusters.persistent` have any effect when constructor arg is used already?

Thanks

**Comment by michael-grunder** [comment, marked as answer]
1. `redis.clusters.cache_slots` uses `PHP`'s `persistent_list` mechanism, which survives multiple requests.  It will live until we invalidate it (e.g. PhpRedis receives a `MOVED` response from the cluster).  It's important to note however, that each PHP-FPM worker will have it's own cache.  So if you have 8 workers, they all need to warm up but at that point, you shouldn't see additional calls to `CLUSTER SLOTS`.
2. `redis.pconnect.pooling_enabled` enables persistent connection pooling.  `RedisCluster` will use the pool if persistent connections are requested in the constructor..
3. `redis.clusters.persistent` is just used when connecting to "named" clusters which can be specified in `php.ini`.  To my knowledge this feature isn't used much.

I created [a little toy script](https://gist.github.com/michael-grunder/a9799c8496f7f01dcab1213391ad5978) that connects to a constructor, echos a value and keeps track of how many times `CLUSTER SLOTS` is called.

You can use php's built-in webserver with an undocumented env variable to watch as the slots get cached in each worker.

```bash
# Start PHP's cli based webserver with 4 workers.
PHP_CLI_SERVER_WORKERS=4 php -S "localhost:9999" -t path/to/count-cluster-slots.php

# Then in another terminal.  You can also use a browser if that's more convenient.
w3m "http://localhost:9999/count-cluster-slots.php?host=localhost&port=7000" -dump
```

As you hit the URL you should see `RedisCluster` executing `CLUSTER SLOTS` once per `PID`.  At that point the PID's `persistent_list` cache is warm and it need not call the command again.

If it continues to call `CLUSTER SLOTS` we'll have to investigate further.

**Comment by gskema** [comment]
Thank you for the test script. I ran it and indeed `CLUSTER SLOTS` is not called continuously. It is sometimes called in `RedisCluster::__construct` when it's not cached but I think that's ok. Sometime it's not cached for first 2-5 requests but them after that it pretty stable? However I still cannot figure out what's going on with the first operation taking twice as long:

```
0.000113, // RedisCluster::__construct (slot cache hit)
0.168065, // GET key1
0.096504, // GET key1
0.090341, // GET key1
```

```
0.1810691, // RedisCluster::__construct (slot cache miss, CLUSTER SLOTS called)
0.1916060, // GET key1 - still double time ?
0.0845618, // GET key1
0.0882408, // GET key1
```

```
0.00009512, // RedisCluster::__construct
0.13961195, // GET key1 - double
0.15950989, // GET key2 - double
0.08325195, // GET key2
0.07949614, // GET key2
```

Sometimes (like <20%) of the time it's not double, feeling like correct slot/node for random key is hit immediately. Other times (~80%) it's around double time taken, like it get MOVED response or something, but the slots are already cached judging by constructor time taken.

**Reply by michael-grunder** [reply]
> I still cannot figure out what's going on with the first operation taking twice as long

The actual connection to the node?  This would also mean that you occasionally don't see the spike when it just so happens to be receiving a key from the node used to cache the slots (done randomly).

**Reply by gskema** [reply]
I will mark this as answered since the slot caching was resolved, but will make a new question about these operation timings
